x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/esql/EsqlDataExtractor.java

```diff
@@ -0,0 +1,166 @@
+        this.isCancelled = false;
+    }
+
+    // TODO: check whether these expressions facilitate injection attacks!
+    private String esqlTimeFilter() {
+        return Strings.format(
+            " | WHERE %s >= \"%s\" AND %s < \"%s\"",
+            timeField,
+            DateUtils.UTC_DATE_TIME_FORMATTER.formatMillis(
+                Math.min(interval.startMs(), org.elasticsearch.common.time.DateUtils.MAX_MILLIS_BEFORE_9999)
+            ),
+            timeField,
+            DateUtils.UTC_DATE_TIME_FORMATTER.formatMillis(
+                Math.min(interval.endMs(), org.elasticsearch.common.time.DateUtils.MAX_MILLIS_BEFORE_9999)
+            )
+        );
+    }
+
+    private String esqlSortByTime() {
+        return Strings.format(" | SORT %s", timeField);
+    }
+
+    private String esqlSummaryStats() {
+        return Strings.format(" | STATS earliest_time=MIN(%s), latest_time=MAX(%s), total_hits=COUNT(*)", timeField, timeField);
+    }
+
+    @Override
+    public DataSummary getSummary() {
+        EsqlQueryRequest request = new EsqlQueryRequest();
+        request.query(datafeed.getEsqlQuery() + esqlTimeFilter() + esqlSummaryStats());
+
+        try (EsqlQueryResponse response = execute(request)) {
+            Iterator<Object> values = response.values().next();
+            String earliestTime = (String) values.next();
+            String latestTime = (String) values.next();
+            long totalHits = (long) values.next();
+            return new DataSummary(
+                earliestTime == null ? null : DateUtils.UTC_DATE_TIME_FORMATTER.parseMillis(earliestTime),
+                latestTime == null ? null : DateUtils.UTC_DATE_TIME_FORMATTER.parseMillis(latestTime),
+                totalHits
+            );
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        return false;
+    }
+
+    @Override
+    public Result next() throws IOException {
+        EsqlQueryRequest request = new EsqlQueryRequest();
+        request.query(datafeed.getEsqlQuery() + esqlTimeFilter() + esqlSortByTime());
+
+        EsqlQueryResponse response = execute(request);
+
+        try (BytesStreamOutput outputStream = new BytesStreamOutput()) {
+            XContentBuilder jsonBuilder = new XContentBuilder(JsonXContent.jsonXContent, outputStream);
+
+            List<ColumnInfo> columns = response.columns();
+            int valueCount = 0;
+            for (Iterator<Iterator<Object>> itValues = response.values(); itValues.hasNext();) {
+                jsonBuilder.startObject();
+                int index = 0;
+                for (Iterator<Object> itValue = itValues.next(); itValue.hasNext();) {
+                    Object value = itValue.next();
+                    if ("date".equals(columns.get(index).type())) {
+                        if (value instanceof String && Strings.isNullOrEmpty((String) value) == false) {
+                            value = DateUtils.UTC_DATE_TIME_FORMATTER.parseMillis((String) value);
```

Comment:
Does ES|QL _always_ output date strings in the same format? If the user has a choice about this then this code might not be generic enough.

Reply:
Yes, I think so, see:
https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java#L145-L148